ls1intum/Artemis
# `General`: Improve authentication of users by storing JWT in the cookie instead of local storage

### Checklist
#### General
#### Server
#### Client

### Motivation and Context
Close #5684 

As detailed [here](https://dev.to/cotter/localstorage-vs-cookies-all-you-need-to-know-about-storing-jwt-tokens-securely-in-the-front-end-15id), it's preferrable to store the authentication token (jwt) in the cookies instead of local storage. It's less prone to XSS attacks. It also allows us to remove a lot of code related to sending/receiving/storing the jwt, as it is all handled by the browser when using cookies. It also gets rid of the need for temporary access tokens to download files.

### Description

To ensure security and as explained in the article above, the cookie set by the server is `secure`, `httpOnly` and `sameSite=Lax`. Same site is set to Lax so that the cookie is sent when users navigate into Artemis. Did not add anti-CSRF tokens as it's not required since we are using a `sameSite=Lax` cookie and for websockets the authentication is sent in the initial HTTP upgrade request, which again is also covered by sameSite.

Also added a check to the websocket handshake request for the cookie/jwt. The handshake is rejected if the jwt is missing or invalid.

The only "functional" change was to change the "api/authenticate" endpoint to set a cookie containing the jwt instead of returning the jwt which the client then stores in localStorage. Because of this change, other changes were necessary:

- Add logout endpoint "api/logout" to "unset" the cookie, as only the server can remove the cookie (not the client/browser)
- Remove all client code storing/checking for/removing/using the jwt from localstorage
- Remove all functionality related to getting/sending/using temporary access tokens to download files
- Update all tests related to setting/receiving/storing the jwt in both the client and the server

### Steps for Testing
Prerequisites:
- 1 Instructor
- 2 Students

Normal login/logout:

1. Log in to Artemis
2. Verify the login works
3. Verify the logout works

Navigate into Artemis:

1. Copy the url of an Artemis download file, or a link to Artemis
2. Verify clicking the link from another page works (e.g. Slack in the browser)

Websockets:

1. Login to Artemis
2. Go to the messages page for a course as a student
3. Login to Artemis on another browser
5. Go to the messages page for the same course as a different student
6. Verify the messages are updated as they are sent

LTI launch:

1. Launch an LTI exercise
7. Verify you are logged in as a new user
8. Verify the logout works

Saml login/logout
1. Login to Artemis using Saml2
2. Verify the login works
3. Verify the logout works

### Review Progress

#### Code Review
#### Manual Tests

### Test Coverage
| Class/File | Branch | Line |
|------------|-------:|-----:|
| JWTFilter.java | 83% | 100% |
| TokenProvider.java | 83% | 99% |
| JwtCookieService.java | 50% | 96% |
| UserJWTController.java | 58% | 88% |
| FileResource.java | 60% | 86% |
| LtiService.java | 90% | 99% |
| auth-jwt.service.ts | 75% | 100% |
| user-route-acess.service.ts | 78% | 89% |
| login.service.ts | 76% |  66% |
| auth-expired.interceptor.ts | 80% |  100% |
| errorhandler.interceptor.ts  | 83% |  100% |
